src/hotspot/share/opto/mulnode.cpp

```diff
@@ -1319,231 +1331,225 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+      jlong hi = ~lo;               // 00007FFF
+      const TypeInteger* t11 = phase->type(in(1)->in(1))->isa_integer(bt);
       if (t11 == nullptr) {
         return this;
       }
       // Does actual value fit inside of mask?
-      if (lo <= t11->_lo && t11->_hi <= hi) {
+      if (lo <= t11->lo_as_long() && t11->hi_as_long() <= hi) {
         return in(1)->in(1);      // Then shifting is a nop
       }
     }
   }
   return this;
 }
 
-//------------------------------Ideal------------------------------------------
-Node *RShiftINode::Ideal(PhaseGVN *phase, bool can_reshape) {
-  // Inputs may be TOP if they are dead.
-  const TypeInt *t1 = phase->type(in(1))->isa_int();
-  if (!t1) return nullptr;        // Left input is an integer
-  const TypeInt *t3;  // type of in(1).in(2)
-  int shift = maskShiftAmount(phase, this, BitsPerJavaInteger);
-  if (shift == 0) {
-    return nullptr;
-  }
-
-  // Check for (x & 0xFF000000) >> 24, whose mask can be made smaller.
-  // Such expressions arise normally from shift chains like (byte)(x >> 24).
-  const Node *mask = in(1);
-  if( mask->Opcode() == Op_AndI &&
-      (t3 = phase->type(mask->in(2))->isa_int()) &&
-      t3->is_con() ) {
-    Node *x = mask->in(1);
-    jint maskbits = t3->get_con();
-    // Convert to "(x >> shift) & (mask >> shift)"
-    Node *shr_nomask = phase->transform( new RShiftINode(mask->in(1), in(2)) );
-    return new AndINode(shr_nomask, phase->intcon( maskbits >> shift));
-  }
-
-  // Check for "(short[i] <<16)>>16" which simply sign-extends
-  const Node *shl = in(1);
-  if( shl->Opcode() != Op_LShiftI ) return nullptr;
-
-  if( shift == 16 &&
-      (t3 = phase->type(shl->in(2))->isa_int()) &&
-      t3->is_con(16) ) {
-    Node *ld = shl->in(1);
-    if( ld->Opcode() == Op_LoadS ) {
-      // Sign extension is just useless here.  Return a RShiftI of zero instead
-      // returning 'ld' directly.  We cannot return an old Node directly as
-      // that is the job of 'Identity' calls and Identity calls only work on
-      // direct inputs ('ld' is an extra Node removed from 'this').  The
-      // combined optimization requires Identity only return direct inputs.
-      set_req_X(1, ld, phase);
-      set_req_X(2, phase->intcon(0), phase);
-      return this;
-    }
-    else if (can_reshape &&
-             ld->Opcode() == Op_LoadUS &&
-             ld->outcnt() == 1 && ld->unique_out() == shl)
-      // Replace zero-extension-load with sign-extension-load
-      return ld->as_Load()->convert_to_signed_load(*phase);
-  }
-
-  // Check for "(byte[i] <<24)>>24" which simply sign-extends
-  if( shift == 24 &&
-      (t3 = phase->type(shl->in(2))->isa_int()) &&
-      t3->is_con(24) ) {
-    Node *ld = shl->in(1);
-    if (ld->Opcode() == Op_LoadB) {
-      // Sign extension is just useless here
-      set_req_X(1, ld, phase);
-      set_req_X(2, phase->intcon(0), phase);
-      return this;
-    }
-  }
-
-  return nullptr;
-}
```

Comment:
This is the same code as above, right? As commented above: it would be good to not move it and reduce the size of the diff.

Reply:
I reordered the changes. Can you take another look?